### ISSUE TITLE
integrate(exp(x**2)*cos(x), x) cannot compute the integral

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#14398]
I don't know if a more simple closed-form exists, but this integral can be computed by rewriting in terms of exp and collecting terms:

```
>>> powsimp(expand(exp(x**2)*cos(x).rewrite(exp)))
exp(x**2 - I*x)/2 + exp(x**2 + I*x)/2
>>> integrate(powsimp(expand(exp(x**2)*cos(x).rewrite(exp))))
sqrt(pi)*exp(1/4)*erfi(x - I/2)/4 + sqrt(pi)*exp(1/4)*erfi(x + I/2)/4
```

Aside from the `rewrite(exp)`, it also would be nice if the powsimp/expand part wasn't necessary. Right now integrate can complete the square in an exponential to result in an erf, but only in a naive way. 

I think there are other issues related to this (I remember seeing it before), but I can't find them at the moment. 

I'm not sure which algorithm this should be fixed in. If meijerg can be improved, that would be great. Some manualintegrate rules would be nice too.

**Comment by asmeurer** [comment]
Wolfram Alpha gives the same answer. So I don't know if a form that doesn't use `I` is possible.

**Comment by sidhantnagpal** [comment]
Would it be helpful if we evaluate the integral `exp(x**2 + I*x)` only, and take the real part of the indefinite integral obtained?
This technique can also be used for evaluation of the summation ([issue](https://github.com/sympy/sympy/issues/14412#issuecomment-371776069)) involving `binomial coefficient` and `cosine`.
That way only one integral/summation is required and the answer can be obtained in a single step by using `(z + conjugate(z))/2` for the real part.

**Comment by normalhuman** [comment]
We don't know if x is real. By default, x can be anywhere in the complex plane, and then `cos(x)` is not the real part of `exp(I*x)`.

**Comment by sidhantnagpal** [comment]
Yes, that is correct (only for real plane), and it would require tweaking for composite terms of `cosine` and `sine`. For example: `exp(x**2)*cos(x)*sin(5*x)*cos(4*x)`.

**Comment by normalhuman** [comment]
Seems a bit like special-casing such functions. I think we'd improve situation in greater generality if [at this stage](https://github.com/sympy/sympy/blob/master/sympy/integrals/meijerint.py#L1855) the rewrite was done not only for hyperbolic trigonometric functions, but also for regular ones. 

Consider the output of meijerint for the much simpler `exp(x)*sin(x)`. It's a monster. 
```
>>> integrate(exp(x)*sin(x), x, meijerg=True)
-2**(3/4)*sqrt(pi)*sqrt(x)*exp(-35*I*pi/8)*besseli(-1/2, sqrt(2)*x*exp_polar(-3*I*pi/4))/8 - 3*2**(3/4)*sqrt(x)*exp(-25*I*pi/8)*besseli(5/2, sqrt(2)*x*exp_polar(-3*I*pi/4))*gamma(-3/4)*gamma(-1/4)/(128*sqrt(pi)) - 2**(3/4)*sqrt(pi)*sqrt(x)*exp(-35*I*pi/8)*besselj(-1/2, sqrt(2)*x*exp_polar(-3*I*pi/4))/8 - 3*2**(3/4)*sqrt(x)*exp(-25*I*pi/8)*besselj(5/2, sqrt(2)*x*exp_polar(-3*I*pi/4))*gamma(-3/4)*gamma(-1/4)/(128*sqrt(pi)) + I*pi*x*besseli(1/2, x*exp_polar(-I*pi/2))*besselj(1/2, x*exp_polar(-I*pi/2))*gamma(1/4)/(16*gamma(5/4)) - 9*2**(1/4)*exp(-35*I*pi/8)*besseli(3/2, sqrt(2)*x*exp_polar(-3*I*pi/4))*gamma(-3/4)*gamma(-1/4)/(128*sqrt(pi)*sqrt(x)) + 9*2**(1/4)*exp(-35*I*pi/8)*besselj(3/2, sqrt(2)*x*exp_polar(-3*I*pi/4))*gamma(-3/4)*gamma(-1/4)/(128*sqrt(pi)*sqrt(x))
```
Compare to what it gives with  rewrite(exp):
```
>>> integrate((exp(x)*sin(x)).rewrite(exp), x, meijerg=True)
-I*(exp(x)*exp(I*x)/2 - I*exp(x)*exp(I*x)/2 - exp(x)*exp(-I*x)/2 - I*exp(x)*exp(-I*x)/2)/2
```
Not optimal either, of course - when we do 
```
>>> integrate((exp(x)*sin(x)), x)
exp(x)*sin(x)/2 - exp(x)*cos(x)/2
```
the answer comes from manualintegrate. But the point is, the second one is way better.

**Comment by normalhuman** [comment]
Scratch that, meijerint cannot handle `exp(x**2 + I*x)` anyway. Issues with conditions for the applicability of integral transform... 

Currently the only routine that handles this integral is heurisch, which makes sense enough.  Its current strategy is to rewrite sin, cos in terms of tan (reducing the number of functions involved), but when cos comes with exp and can be absorbed there, that would probably be a better rewrite...

**Comment by asmeurer** [comment]
Not even if the square is completed?

**Comment by normalhuman** [comment]
With completed square it does. That's pretty fragile.
